### ISSUE TITLE
SSL issue pulling soundcloud audio

### ISSUE THREAD

**Issue by albill** [snarfed/huffduff-video#24]
When I try to pull an audio file from soundcloud, I'm seeing the following SSL handshake error:

huffduff-video
Fetching https://soundcloud.com/motherboard/cryptoparty-in-harlem ...

ERROR: Unable to download JSON metadata: (caused by URLError(SSLError(1, '_ssl.c:493: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure'),))

**Comment by snarfed** [comment]
thanks for reporting, and sorry for the trouble! will look.

**Comment by albill** [comment]
Whatever you did has fixed the issue.

**Comment by albill** [comment]
aaaaaand...it's back. :-)

Fetching https://soundcloud.com/panoply/elizabeth-kolbert-we-have-locked-in-centuries-of-climate-change ...

ERROR: Unable to download JSON metadata: (caused by URLError(SSLError(1, '_ssl.c:493: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure'),))

**Comment by albill** [comment]
Isn't this just the need to update the youtube-dl script?

**Comment by snarfed** [comment]
@albill often, yes! not this time, though. i tried that originally, and again just now - huffduff-video is currently on youtube_dl-2017.2.28 - and no luck either time. whatever fixed it originally, it wasn't me. :/

**Comment by snarfed** [comment]
rg3/youtube-dl#12280 says this is because soundcloud requires SNI (!), and support for that is dependent on openssl version, system cert bundle, etc. will investigate more.

**Comment by snarfed** [comment]
data points:
* `youtube-dl https://soundcloud.com/panoply/elizabeth-kolbert-we-have-locked-in-centuries-of-climate-change` works fine. so apache/mod_wsgi may be using a different openssl or python version than command line youtube-dl.
* `python2 -c 'import ssl; print(ssl.OPENSSL_VERSION)'` (recommended in rg3/youtube-dl#12280) complains `AttributeError: 'module' object has no attribute 'OPENSSL_VERSION'` (!)
* `openssl s_client -connect api.soundcloud.com:443 -servername api.soundcloud.com` verifies the cert ok:
    ```
    subject=/OU=Domain Control Validated/CN=*.soundcloud.com
    issuer=/C=BE/O=GlobalSign nv-sa/CN=GlobalSign Domain Validation CA - SHA256 - G2
    ...
    Verify return code: 0 (ok)
    ```

**Comment by snarfed** [comment]
```sh
$ python -V
Python 2.7.12
$ openssl version
OpenSSL 1.0.1k-fips 8 Jan 2015
```

**Comment by snarfed** [comment]
interestingly, `python2` on the server runs python 2.6, not 2.7. not sure what apache's mod_wsgi is using.

```sh
$ python2 -V
Python 2.6.9

$ python2 -c 'import ssl; print(ssl.OPENSSL_VERSION)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'module' object has no attribute 'OPENSSL_VERSION'

$ python -V
Python 2.7.12

$ python -c 'import ssl; print(ssl.OPENSSL_VERSION)'
OpenSSL 1.0.1k-fips 8 Jan 2015
```

**Comment by snarfed** [comment]
confirmed that apache's mod_wsgi is using python 2.6.9. that's the problem. :/ hopefully rebuilding it with python 2.7.12 will fix this.

**Comment by snarfed** [comment]
worked! soundcloud works now. thanks for your patience @albill.

**Comment by albill** [comment]
Confirmed! Thanks for the work.

**Comment by snarfed** [comment]
looks like it's back again. :/

**Comment by snarfed** [comment]
the fix didn't regress, we're still on Python 2.7.12 and OpenSSL 1.0.1k-fips, so something else is wrong.

we're now getting an HTTP 401 Unauthorized, as opposed to the SSLError connection failed from before. i suspect soundcloud is just blocking EC2 IP addresses, or youtube-dl's current soundcloud client id. :/

**Comment by snarfed** [comment]
woo youtube-dl updated their soundcloud client id in version 2017.03.07, which fixed it! closing.
